### PR TITLE
validate manual server URLs before entry, leniently

### PR DESCRIPTION
**Pull request by trevj**
Use https://www.npmjs.com/package/jsonic to parse server management URLs. This helps us:
- ignore missing outer squiggly brackets
- ignore missing quotes
- accept "weird" quotes and accept a mix of quote styles

I preserved the existing behaviour of ignoring any content before and after the squiggly brackets, if present.

I also disabled the DONE button until the contents of the box are valid; we don't have to do this but, IIRC, the only reason we haven't done this was lack of communication between the Polymer template and the TypeScript code.

@sandrigo FYI

### CODE REVIEW

**Comment by trevj** [comment]
BTW @alalamav although I've branched from master this will be an easy merge when multiple servers is merged to master.

File: src/server_manager/web_app/management_urls.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  const indexOfFirstBrace = userInput.indexOf('{');
+  if (indexOfFirstBrace >= 0) {
+    userInput = userInput.substring(indexOfFirstBrace);
+  }
+  const indexOfLastBrace = userInput.lastIndexOf('}');
+  if (indexOfLastBrace >= 0) {
+    userInput = userInput.substring(0, indexOfLastBrace + 1);
+  }
+
+  const config = jsonic(userInput) as ManualServerConfig;
```
[the remark this answers is not in the document]
Comment (trevj) [reply src/server_manager/web_app/management_urls.ts]: No - but the code immediately after will "verify" the object has the required fields.

File: src/server_manager/web_app/management_urls.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import * as jsonic from 'jsonic';
+
+import {ManualServerConfig} from '../model/server';
+
+// Parses a server management URL (generated by the server install script) leniently.
+// TODO: ignore case of key names
+// TODO: check value types
+export function parseManualServerConfig(userInput: string): ManualServerConfig {
+  // Remove anything before and after the first and last braces, if found.
+  const indexOfFirstBrace = userInput.indexOf('{');
```
[the remark this answers is not in the document]
Comment (trevj) [reply src/server_manager/web_app/management_urls.ts]: This is the code that actually removes the content before and after the braces :-)

[the remark this answers is not in the document]
Comment (alalamav) [reply src/server_manager/web_app/management_urls.ts]: 😅

File: src/server_manager/web_app/app.ts [hunk 3]
```diff
@@ -120,11 +121,23 @@ export class App {
...
+        parseManualServerConfig(event.detail.userInput);
+      } catch (e) {
+        isValid = false;
+      }
+      const manualServerEntryEl = appRoot.getManualServerEntry();
+      manualServerEntryEl.enableDoneButton = isValid;
+    });
+
     appRoot.addEventListener('ManualServerEntered', (event: PolymerEvent) => {
-      const userInputConfig =
```
Comment (fortuna) [inline src/server_manager/web_app/app.ts]: It seems we need to restore this.

**Pull request merged**